src/main/java/com/hivemq/mqtt/services/PublishPollServiceImpl.java

```diff
@@ -222,7 +223,12 @@ public void onSuccess(final ImmutableList<MessageWithID> messages) {
                 for (int i = 0, messagesSize = messages.size(); i < messagesSize; i++) {
                     final MessageWithID message = messages.get(i);
                     final FreePacketIdRanges freePacketIdRanges = clientConnection.getFreePacketIdRanges();
-                    freePacketIdRanges.takeIfAvailable(message.getPacketIdentifier());
+                    try {
+                        freePacketIdRanges.takeSpecificId(message.getPacketIdentifier());
+                    } catch (final MessageIdUnavailableException e) {
+                        log.warn("The desired packet ID was not available when polling inflight messages: {}",
+                                e.getMessage());
+                    }
```

Comment:
The previous implementation dropped the message here and returned. Seems safer to me to avoid continuing with duplicate packet IDs here. What do you think?

Reply:
From what I see it happened only on NoMessageIdAvailableException which is different from MessageIdUnavailableException which we throw now; also the behavior of the previous code was very weird - it took the packet ID checked if it matched with the current packet's id and then returned, so the taken id was never used... Abdullah suggested to get rid of this behavior altoghether. I think that MessageIdUnavailableException is not really a reason to drop the message since it does not mean that queue is full or we ran out of ids, it means that we somehow managed to trigger an unspecified behavior. It is useful for debugging purpose to have this log but I would not drop packets in such case tbh

Reply:
Thanks, makes sense to me. Also had an unrelated chat with Lukas that lead us to packet IDs again, and it seems like the whole implementation schema is a bit of a mix of cluster-like behavior (where the connected node dictates the packet IDs, which we want to remove) and previous attempts to send messages directly without enqueueing them. So this is fine. 👍